decred/decred/dcr/account.py

```diff
@@ -831,19 +1073,180 @@ def calcBalance(self, tipHeight=None):
-                ticketVal += utxo.satoshis
-                self.tickets.append(utxo.txid)
+        uT = self.unspentTickets()
+        ticketCount = len(uT)
+        ticketVal = sum(t.satoshis for t in uT.values())
         self.stakeStats = TicketStats(ticketCount, ticketVal)
 
+    def spentTicketInfo(self, txid):
+        """
+        It's possible that when retrieving the tinfo from dcrdata that it is not
+        ready. Log a message and return None if so. Return the TicketInfo if found.
+
+        Args:
+            txid (str): The ticket to find info for.
+
+        Returns:
+            TicketInfo: The ticket info if found or None.
+        """
+        tinfo = self.blockchain.ticketInfo(txid)
+        if tinfo.vote or tinfo.revocation:
+            return tinfo
+        log.debug("ticket's tinfo not found on the blockchain {}".format(txid))
+
+    def spendTicket(self, tx):
+        """
+        Update a ticket that is already in the ticketDB. Sends a
+            ui.SPENT_TICKETS_SIGNAL on completion.
+
+        Args:
+            tx (msgtx.MsgTx): the vote or transaction that spends a ticket in
+                the ticketDB.
+        """
+        isVote = txscript.isSSGen(tx)
+        idx = 1 if isVote else 0
+        ticketTxid = tx.txIn[idx].previousOutPoint.txid()
+        if ticketTxid not in self.ticketDB:
+            raise Exception(
+                "spending tx came for ticket that was not found in the database: {}".format(
+                    ticketTxid
+                )
+            )
+        # sometimes this data cannot be fetched yet by dcrdata
+        tinfo = self.spentTicketInfo(ticketTxid)
+        if not tinfo:
+            tinfo = self.blockchain.ticketInfoForSpendingTx(tx.txid(), self.net)
+
+        ticket = self.ticketDB[ticketTxid]
+        ticket.tinfo = tinfo
+        self.ticketDB[ticketTxid] = ticket
+
+        self.signals.spentTickets()
+
+    def updateSpentTickets(self):
+        """
+        Update voted and revoked tickets. ui.SPENT_TICKETS_SIGNAL is called if
+        there was an update.
+        """
+        # Get current unspent tickets.
+        mempoolTickets = {
+            k: UTXO.ticketFromTx(v, self.net)
+            for k, v in self.mempool.items()
+            if v.isTicket()
+        }
+        activeTickets = {
+            utxo.txid: utxo for utxo in self.utxos.values() if utxo.isTicket()
+        }
+        activeTickets.update(mempoolTickets)
+        # Get tickets as stored in the database.
+        unspentTickets, votedTickets, revokedTickets = self.sortedTickets()
+
+        # Only update if we have reason to believe ticket statuses have changed.
+        if set(activeTickets.keys()) == set(unspentTickets.keys()):
+            if all(
+                (
+                    activeTickets[k].tinfo.status == unspentTickets[k].tinfo.status
+                    for k in unspentTickets.keys()
+                )
+            ):
+                return
```

Comment:
I made some suggestions to make this more efficient at 

https://github.com/JoeGruffins/tinydecred/compare/tickets...buck54321:jg?expand=1

See what you think.

Reply:
It looks like this will not search for spent tickets we have never seen. Is that ok? I suppose a seperate search could be connected to adding a new stakepool. That does sound better to me actually.

Reply:
Let me know if it looks any better now.